### PR TITLE
ENH: Use Kahan summation and Welfords method to calculate rolling var and std

### LINKED ISSUE #37051: DOC/PERF: Decide how to handle floating point artifacts during rolling calculations
Currently we have a check here that artificially handles a numerical precision issue in `rolling.var` and `rolling.std` where our rolling variance calculation is carrying forward floating point artifacts. Ideally we should be using a more numerically stable algorithm (maybe Kahan summation) so this check isn't so arbitrary. 

https://github.com/pandas-dev/pandas/blob/601eff1768380313ec452ba8677cc7e825706796/pandas/_libs/window/aggregations.pyx#L305

### PR DESCRIPTION
**Pull request by phofl**
- [x] xref #37051
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

As suggested by @mroeschke Kahan summation fixes the numerical problems. Additionally I used Welfords Method to calculate ``ssqdm``, because previously the tests I have added would return

```
0             NaN
1    3.500000e+34
2    3.000000e+34
3    3.000000e+34
4    3.000000e+34
5    3.000000e+34
6    3.000000e+34
7    3.000000e+34
8    3.000000e+34
9    3.000000e+34
dtype: float64
```

for ``var()``. I am running the asv and will post the results when available

### CODE REVIEW

**Comment by jreback** [comment]
great thanks @phofl

**Comment by phofl** [comment]
Unfortunately I chose a bad example, which did miss one bug. Fixed it now and added the corresponding test.

**Comment by jreback** [comment]
xref https://github.com/pandas-dev/pandas/pull/6817

i guess this was there a long time ago, but i dont' think had enough tests to lock it down.

**Comment by jreback** [comment]
and maybe some examples from here: https://github.com/pandas-dev/pandas/issues/6929 (though that's obviously a separate issue)

**Comment by phofl** [comment]
Yes planned to Look into this in the future. Maybe we can improve this in a similar way. 

Delta**2 was the problem with the modified version. Switching to regular welford fixes this

**Comment by phofl** [comment]
Looks like this will only help with large numbers.

**Comment by phofl** [comment]
Interestingly 

BENCHMARKS NOT SIGNIFICANTLY CHANGED.

**Comment by jreback** [comment]
can you merge master, ping on green

**Comment by phofl** [comment]
@jreback green

File: pandas/_libs/window/aggregations.pyx [hunk 1]
```diff
@@ -353,7 +362,8 @@ def roll_var(ndarray[float64_t] values, ndarray[int64_t] start,
     Numerically stable implementation using Welford's method.
     """
     cdef:
-        float64_t mean_x = 0, ssqdm_x = 0, nobs = 0,
+        float64_t mean_x = 0, ssqdm_x = 0, nobs = 0, compensation_add = 0,
+        float64_t compensation_remove = 0,
```
Comment (jreback) [inline pandas/_libs/window/aggregations.pyx:366]: isn't there a line in this func that you need to remove 

eg the < 1e-14

?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
@phofl this PR doesn't the close issue? can u show an example of when

**Comment by phofl** [comment]
Yeah thought so too initially, because I was not able to construct a counter example. But our docstrings do the job:
If we remove the line:
```
s = pd.Series([5, 5, 6, 7, 5, 5, 5])
print(s.rolling(3).var())
```

Returns

```
0             NaN
1             NaN
2    3.333333e-01
3    1.000000e+00
4    1.000000e+00
5    1.333333e+00
6    6.661338e-16
dtype: float64
```

Seems like Kahan summation and Welfords method only help for large numbers. Issues with numbers like ``1/3`` like here can't be fixed with that.

File: pandas/_libs/window/aggregations.pyx (hunk 1 above)
Comment (phofl) [reply pandas/_libs/window/aggregations.pyx:366]: See below

**Comment by jreback** [comment]
> Yeah thought so too initially, because I was not able to construct a counter example. But our docstrings do the job:
> If we remove the line:
> 
> ```
> s = pd.Series([5, 5, 6, 7, 5, 5, 5])
> print(s.rolling(3).var())
> ```
> 
> Returns
> 
> ```
> 0             NaN
> 1             NaN
> 2    3.333333e-01
> 3    1.000000e+00
> 4    1.000000e+00
> 5    1.333333e+00
> 6    6.661338e-16
> dtype: float64
> ```
> 
> Seems like Kahan summation and Welfords method only help for large numbers. Issues with numbers like `1/3` like here can't be fixed with that. Rounding issues after multiplication cause there problems.

ok its prob worth adding an xfail test for that one. (followon ok)

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -192,6 +192,7 @@ Other enhancements
 - Added methods :meth:`IntegerArray.prod`, :meth:`IntegerArray.min`, and :meth:`IntegerArray.max` (:issue:`33790`)
 - Where possible :meth:`RangeIndex.difference` and :meth:`RangeIndex.symmetric_difference` will return :class:`RangeIndex` instead of :class:`Int64Index` (:issue:`36564`)
 - Added :meth:`Rolling.sem()` and :meth:`Expanding.sem()` to compute the standard error of mean (:issue:`26476`).
+- :meth:`Rolling.var()` and :meth:`Rolling.std()` use Kahan summation and Welfords Method to avoid numerical issues (:issue:`37051`)
```
Comment (jreback) [inline doc/source/whatsnew/v1.2.0.rst:195]: this is not fully true, but its better so ok.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl

**Comment by phofl** [comment]
@jreback with the line < 1e-14 this test would not fail. I could add a test which passes, but would fail, if somebody removes the line without fixing the underlying problem?

**Pull request merged**